web/docs/guides/websockets.md

```diff
@@ -0,0 +1,129 @@
+---
+title: WebSockets
+---
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
+# WebSocket support
+Wasp provides a fully integrated WebSocket experience by utilizing [Socket.IO](https://socket.io/) on the client and server. We handle making sure your URLs are correctly setup, CORS is enabled, and provide a useful `useSocket` abstraction for use in React components.
+
+## Enable in Wasp file
+Here, we specify we are using WebSockets by adding `webSocket` to our `app` and providing the required `fn`. You can optionally specify a list of `entities` for use in your context (like in operations).
+
+```wasp title=todoApp.wasp
+app todoApp {
+  // ...
+
+  webSocket: {
+    fn: import { webSocketFn } from "@server/webSocket.js"
+    // Optional: entities: [...]
+  },,
+}
+```
+
+## Optional: Add shared types for client and server use
+If TypeScript is your thing, you can add shared interface definitions to use on both the client and server. This allows you to type all the messages you send and receive!
+
+```ts title=src/shared/webSocket.ts
+import { WaspSocketData } from '@wasp/universal/types'
+
+export interface ServerToClientEvents {
+  chatMessage: (msg: { id: string, username: string, text: string }) => void;
+}
+
+export interface ClientToServerEvents {
+  chatMessage: (msg: string) => void;
+}
+
+export interface InterServerEvents { }
+
+// Data that is attached to the socket.
+// NOTE: Wasp automatically injects the JWT into the connection,
+// and if present/valid, the server adds a user to the socket.
+export interface SocketData extends WaspSocketData { }
+```
+
+## Add server support
+On the server, you will get `io` argument and `context` for your WebSocket function. You can use this `io` object to register callbacks for all the regular Socket.IO events.
+
+```ts title=src/server/webSocket.ts
+import { v4 as uuidv4 } from 'uuid'
+import { WebSocketDefinition } from '@wasp/webSocket'
+import { ServerToClientEvents, ClientToServerEvents, InterServerEvents, SocketData } from '../shared/webSocket'
+
+export const webSocketFn: WebSocketDefinition<
+  ClientToServerEvents,
+  ServerToClientEvents,
+  InterServerEvents,
+  SocketData
+> = (io, context) => {
+  io.on('connection', (socket) => {
+    const username = socket.data.user?.email || socket.data.user?.username || 'unknown'
+    console.log('a user connected: ', username)
+
+    socket.on('chatMessage', async (msg) => {
+      console.log('message: ', msg)
+      io.emit('chatMessage', { id: uuidv4(), username, text: msg })
+    })
+  })
+}
+```
+
+## Add client support
+Client access to WebSockets is provided by the `useSocket` abstraction. It returns:
```

Comment:
What if they want to use `socket` directly, without `useSocket`, should we allow them that? What if they want to make connection conditional, and have it connected only sometimes / when used? Hm this sounds like reasonable request, have socket connected only when used. I am not saying we have to support all this, but would love to explore use cases a bit.

Ok let's imagine this: I have an app which has a chat in it. Chat is used only sometimes. So with this PR, all instances of my Wasp app will maintain active connection, even though possibly only 1% of them is using the socket? Could that be an issue? Is there a limit to number of socket connections that can even be opened?

Reply:
Ok, so I pulled `refreshAuthToken` up to the module level, exported it and `socket`, and added an `autoConnect` option. So this effectively does 1 and 2, and they can now support your use case by setting `autoConnect: false` and then manually calling `socket.connect()`/`socket.disconnect()` in the component that needs it 👍🏻

Reply:
But by default, we will still just have `autoConnect` be `true`

Reply:
oh ok that is pretty cool now, sounds good!

Could you elaborate a bit more on the clustering? I don't know the problem you are talking about, maybe you can explain it to me in a bit more detail here? Sounds to me like we should create an issue for this so we can track it in GH?

Reply:
I have a note in the code with a link, but essentially the servers will need to talk to each other if you want all messages to go to everyone. Additionally, if using long polling, we would need sticky session.

For example:
- https://socket.io/docs/v4/using-multiple-nodes
- https://socket.io/docs/v4/cluster-adapter/

If it needs a full issue just let me know 👍🏻 But as of now, this works for number of server containers = 1, not > 1.

Reply:
Ok sounds good -> please create a full Github issue for this, so we can track it!

Reply:
https://github.com/wasp-lang/wasp/issues/1228